FeedStoreChallenge/CoreDataFeedStore.swift

```diff
@@ -29,14 +29,99 @@ public final class CoreDataFeedStore: FeedStore {
 	}
 
 	public func retrieve(completion: @escaping RetrievalCompletion) {
-		fatalError("Must be implemented")
+		perform { context in
+			do {
+				guard let feed = try self.fetchFirstManagedFeed(),
+				      let managedFeedImages = feed.managedFeedImages else {
+					return completion(.empty)
+				}
+				let feedImages = managedFeedImages.compactMap(LocalFeedImage.init(managedFeedImage:))
+				completion(.found(feed: feedImages, timestamp: feed.timestamp!))
+			} catch {
+				completion(.failure(error))
+			}
+		}
 	}
 
 	public func insert(_ feed: [LocalFeedImage], timestamp: Date, completion: @escaping InsertionCompletion) {
-		fatalError("Must be implemented")
+		perform { context in
+
+			let managedFeed = self.createManagedFeed(from: feed)
+			managedFeed.timestamp = timestamp
+
+			do {
+				try context.execute(self.makeDeleteManagedFeedRequest())
+				try context.save()
+				completion(nil)
+			} catch {
+				completion(error)
+				context.rollback()
+			}
+		}
 	}
 
 	public func deleteCachedFeed(completion: @escaping DeletionCompletion) {
-		fatalError("Must be implemented")
+		perform { context in
+			do {
+				try context.execute(self.makeDeleteManagedFeedRequest())
+				completion(nil)
+			} catch {
+				completion(error)
+			}
+		}
+	}
+
+	//MARK: - Private Methods
+
+	private func makeDeleteManagedFeedRequest() -> NSBatchDeleteRequest {
+		NSBatchDeleteRequest(fetchRequest: ManagedFeed.fetchRequest())
+	}
+
+	private func perform(_ action: @escaping (NSManagedObjectContext) -> Void) {
+		let context = self.context
+		context.perform {
+			action(context)
+		}
+	}
+
+	private func fetchFirstManagedFeed() throws -> ManagedFeed? {
+		let request: NSFetchRequest<ManagedFeed> = ManagedFeed.fetchRequest()
+		return try context.fetch(request).first
+	}
+
+	private func createManagedFeedImage(from feedImage: LocalFeedImage) -> ManagedFeedImage {
+		let managedItem = ManagedFeedImage(context: context)
+		managedItem.id = feedImage.id
+		managedItem.imageDescription = feedImage.description
+		managedItem.location = feedImage.location
+		managedItem.url = feedImage.url
+		return managedItem
+	}
+
+	private func createManagedFeed(from feed: [LocalFeedImage]) -> ManagedFeed {
+		let managedFeed = ManagedFeed(context: context)
+		let managedItems = feed.map(createManagedFeedImage(from:))
+		let managedItemsSet = NSOrderedSet(array: managedItems)
+		managedFeed.addToItems(managedItemsSet)
+		return managedFeed
+	}
+}
+
+private extension ManagedFeed {
+	var managedFeedImages: [ManagedFeedImage]? {
```

Comment:
There's no need to return an `optional` array of `ManagedFeedImages` since a `ManagedFeed` with no images would still have a perfectly valid empty `[]` array of images.
```suggestion
	var managedFeedImages: [ManagedFeedImage] {
```